admin/server/projects.go

```diff
@@ -974,6 +974,228 @@ func (s *Server) GetCloneCredentials(ctx context.Context, req *adminv1.GetCloneC
 	}, nil
 }
 
+func (s *Server) RequestProjectAccess(ctx context.Context, req *adminv1.RequestProjectAccessRequest) (*adminv1.RequestProjectAccessResponse, error) {
+	observability.AddRequestAttributes(ctx,
+		attribute.String("args.org", req.Organization),
+		attribute.String("args.project", req.Project),
+	)
+
+	proj, err := s.admin.DB.FindProjectByName(ctx, req.Organization, req.Project)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	claims := auth.GetClaims(ctx)
+	if claims.ProjectPermissions(ctx, proj.OrganizationID, proj.ID).ReadProject {
+		return nil, status.Error(codes.InvalidArgument, "already have access to project")
+	}
+
+	if claims.OwnerType() != auth.OwnerTypeUser {
+		return nil, status.Error(codes.InvalidArgument, "only users can request access")
+	}
+
+	user, err := s.admin.DB.FindUser(ctx, claims.OwnerID())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	org, err := s.admin.DB.FindOrganization(ctx, proj.OrganizationID)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	// TODO: quotas
+
+	existing, err := s.admin.DB.FindProjectAccessRequest(ctx, proj.ID, user.Email)
+	if err != nil && !errors.Is(err, database.ErrNotFound) {
+		return nil, err
+	}
+	if existing != nil {
+		return nil, status.Error(codes.AlreadyExists, "have already requested access to project")
+	}
+
+	accessReq, err := s.admin.DB.InsertProjectAccessRequest(ctx, &database.InsertProjectAccessRequestOptions{
+		Email:     user.Email,
+		ProjectID: proj.ID,
+	})
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	adminRole, err := s.admin.DB.FindOrganizationRole(ctx, "admin")
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	admins, err := s.admin.DB.FindOrganizationMemberUsersByRole(ctx, proj.OrganizationID, adminRole.ID)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
```

Comment:
Instead of hard-coding the role name, It would be better to search for members with the `ManageOrgMembers` permission.

Reply:
Updated